Add unique_id to the statistics component

ThomDietrich opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add optional unique_id to the statistics component.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue:
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/20542

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

ThomDietrich commented:
@emontnemery fyi, ready to merge

emontnemery commented:
Please add tests. The test can for example make sure the entity is added to the entity registry.

ThomDietrich commented:
Isn't that a bit overkill? At this point I am testing the workings of an HA core functionality for SensorEntity within one specific sensors code...
Happy to add if you still say yes.

emontnemery commented:
You're not testing core functionality of Home Assistant, you're testing that the `unique_id` configuration parameter of the statistics sensor works as intended.

It doesn't have to be complicated, just do:
```
from homeassistant.components import sensor
from homeassistant.helpers import entity_registry as er

...

entity_registry = er.async_get(hass)
assert entity_registry .async_get_entity_id(sensor.domain, DOMAIN, "veryunique")
```

emontnemery commented on homeassistant/components/statistics/sensor.py:
```diff
@@ -125,6 +125,7 @@ def __init__(
         self._entity_id = entity_id
         self.is_binary = self._entity_id.split(".")[0] == "binary_sensor"
         self._name = name
+        self._unique_id = f"{self._entity_id.replace('.', '_')}__statistics"
```
Do we really want to use the source sensor's entity id as unique id, it means the unique_id is not stable if the source entity's entity_id is changed. There's nothing guaranteeing the source entity's ID is stable and not changed by the user.

ThomDietrich replied:
Kind of the point. In https://github.com/home-assistant/core/pull/49960 I would add the selected characteristic to the mix.

When the user changes the source entity (or the characteristic) the resulting statistics sensor tracks a new measure and should for example not continue with old measurements in the recorder. So far this made sense to me.

Given, the user could rename the entity_id of the source sensor. I would be good to use the unique_id of the source entity. Sound better?

emontnemery replied:
We mostly don't let the user specify a `unique_id` for integrations where users manually configure entities.
One exception is `automation`.

What problem do you really intend to solve by adding statistics sensors to the entity registry?

ThomDietrich replied:
Generally I thought it's desired to have a unique_ud for every entity. Secondly: 
![image](https://user-images.githubusercontent.com/2870104/140772015-905e01de-8580-4002-8633-2951fff5e2c1.png)
I want to be able to assign to sensor to a room etc.

emontnemery replied:
I'd suggest to mark this as draft and finish off #49960 first.

ThomDietrich commented:
Another reason to implement this.
Related to #49960: When changing the configured characteristic, the recorder will continue the same time series, which if you change from "mean" to "count" or "variance" really doesn't make any sense.

![image](https://user-images.githubusercontent.com/2870104/141678109-eb106d92-bfef-4aa4-819d-d869ee1d5a2b.png)

ThomDietrich commented:
@emontnemery can we pick this up again? I will summarize:

The statistics component is now in a breaking change state. We should use that opportunity to also introduce the unique_id, if we agree that is a good step.

* My understanding was that sooner or later all integrations should introduce unique_ids
* One reason for introducing a unique_id is the graph example shown above: Switching between characteristics does otherwise not start a new recorder history
* I suggested a unique_id build from entity_id and state_characteristic, however there are edge cases
* Alternative approach would be configuration option to manually set unique_id, possibly in combination with the default from the last point

emontnemery commented:
> an optional configuration option to manually set unique_id

Let's do that, without any default.
That's in line with the `unique_id` configuration offered by template entities, including template sensor: https://www.home-assistant.io/integrations/template/#all-sensor-binary-sensor-number-select-entities

ThomDietrich committed:
Implement optional manually defined uniqueid

ThomDietrich commented:
@emontnemery I might have gone with the default additionally but either way is great.

PR ready. Changed to be purely optional parameter and added a test case as requested. Documentation PR exists.

emontnemery commented:
The test problem is because the entity registry is not mocked, so it writes its settings.
This is not wanted, i think the easiest way is to simply refactor the test to be a modern pytest test, not part of the `TestStatisticsSensor` class.

If you do that, and the test uses the `hass` mock, the entity registry will be mocked for you.

The test would then be:
```py
async def test_unique_id(hass):
    ...
```

ThomDietrich committed:
Fix test case via mocked environment

ThomDietrich committed:
Merge remote-tracking branch 'upstream/dev' into add_statistics_uniqueid

ThomDietrich commented:
@emontnemery thanks for the hint. Ready.

> A follow-up PR getting rid of the TestStatisticsSensor would be welcome.

👍

emontnemery reviewed (approved):
Thanks, @ThomDietrich 👍
